### ISSUE TITLE
CPython 3.13 warns on FutureRef._evaluate/typing._eval_type not passing `type_params`

### ISSUE THREAD

**Issue by vfazio** [pydantic/pydantic#9613]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

See upstream pull request https://github.com/python/cpython/pull/118695

And discussion: https://github.com/python/cpython/issues/118418

The current call pattern will be disallowed starting in CPython 3.15 based on the warning.

### Python, Pydantic & OS Version

```Text
Python 3.13
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by vfazio** [comment]
@sydney-runkle FYI, semi related to #9607

**Comment by musicinmybrain** [comment]
Note that this can be reproduced with:

```
$ gh repo clone pydantic/pydantic
$ cd pydantic
$ python3.13 -m venv _e
$ . _e/bin/activate
(_e) $ UNSAFE_PYO3_SKIP_VERSION_CHECK=1 pip install -e .
(_e) $ pip install cloudpickle coverage[toml] dirty-equals eval-type-backport pytest pytest-mock pytest-pretty pytest-examples faker pytest-benchmark pytest-codspeed packaging
(_e) $ python -m pytest
```

```
==================================================================================================== ERRORS ====================================================================================================
________________________________________________________________________ ERROR collecting tests/benchmarks/test_discriminated_unions.py ________________________________________________________________________
tests/benchmarks/test_discriminated_unions.py:11: in <module>
    class NestedState(BaseModel):
pydantic/_internal/_model_construction.py:200: in __new__
    set_model_fields(cls, bases, config_wrapper, types_namespace)
pydantic/_internal/_model_construction.py:469: in set_model_fields
    fields, class_vars = collect_model_fields(cls, bases, config_wrapper, types_namespace, typevars_map=typevars_map)
pydantic/_internal/_fields.py:132: in collect_model_fields
    type_hints = get_cls_type_hints_lenient(cls, types_namespace)
pydantic/_internal/_typing_extra.py:231: in get_cls_type_hints_lenient
    hints[name] = eval_type_lenient(value, globalns, localns)
pydantic/_internal/_typing_extra.py:243: in eval_type_lenient
    return eval_type_backport(value, globalns, localns)
pydantic/_internal/_typing_extra.py:259: in eval_type_backport
    return typing._eval_type(  # type: ignore
/usr/lib64/python3.13/typing.py:470: in _eval_type
    _deprecation_warning_for_no_type_params_passed("typing._eval_type")
/usr/lib64/python3.13/typing.py:450: in _deprecation_warning_for_no_type_params_passed
    warnings.warn(depr_message, category=DeprecationWarning, stacklevel=3)
E   DeprecationWarning: Failing to pass a value to the 'type_params' parameter of 'typing._eval_type' is deprecated, as it leads to incorrect behaviour when calling typing._eval_type on a stringified annotation that references a PEP 695 type parameter. It will be disallowed in Python 3.15.
[…]
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 17 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Results (2.24s):
        17 errors
```

**Comment by musicinmybrain** [comment]
Note also that https://github.com/pydantic/pydantic/issues/9613#issuecomment-2170428152, above, reproduces this with the latest Pydantic v2, so this is not only a Pydantic v1 issue.

**Comment by samuelcolvin** [comment]
Are you sure you mean V1? The errors seem to relate to PyO3 which is only used with V2.

**Comment by vfazio** [comment]
Anything calling `typing._eval_type` or `typing._evaluate` will throw these deprecation warnings i believe so maybe both versions?

**Comment by samuelcolvin** [comment]
yup, makes sense. Ideally we should find a way to remove their use, at least from V2.

In V1 we can do something hackier.

**Comment by sydney-runkle** [comment]
@vfazio,

Are you able to repro this on `main`? Do we need to leave this open as we look to explicitly support 3.13 for v2?

**Comment by vfazio** [comment]
> @vfazio,
> 
> Are you able to repro this on `main`? Do we need to leave this open as we look to explicitly support 3.13 for v2?

@sydney-runkle It looks like @musicinmybrain tested this on `main` https://github.com/pydantic/pydantic/issues/9613#issuecomment-2170428152. If you want me to reproduce, I can set some time aside maybe tomorrow.

**Comment by sydney-runkle** [comment]
All good, let's leave this open then, thanks @musicinmybrain!

**Comment by sydney-runkle** [comment]
@vfazio,

Any interest in helping us with another v1 patch here? I've got a draft PR up to start adding support for 3.13 explicitly in v2, but we'll have to do this `_eval_type` fix in the v1 code...

**Comment by vfazio** [comment]
Hah, I thought we weren't making changes in v1 to accommodate 3.13. Is there no way to not call into v1 apis for this functionality?

**Comment by sydney-runkle** [comment]
Sigh... you and me both. I'll look into this in depth in the morning.

**Comment by sydney-runkle** [comment]
@vfazio,

Since this is such a small change, we're going to roll with it. Hopefully nothing else is needed in V1.

**Comment by sydney-runkle** [comment]
Closing given that I merged the above, this should be fixed in 2.8!
